base/collection.go

```diff
@@ -0,0 +1,445 @@
+		username, password, _ := spec.Auth.GetCredentials()
+		authenticator = gocb.PasswordAuthenticator{
+			Username: username,
+			Password: password,
+		}
+	}
+
+	// Timeouts
+	var timeoutsConfig gocb.TimeoutsConfig
+	if spec.BucketOpTimeout != nil {
+		timeoutsConfig.KVTimeout = *spec.BucketOpTimeout
+		timeoutsConfig.ManagementTimeout = *spec.BucketOpTimeout
+		timeoutsConfig.ConnectTimeout = *spec.BucketOpTimeout
+	}
+	timeoutsConfig.QueryTimeout = spec.GetViewQueryTimeout()
+	timeoutsConfig.ViewTimeout = spec.GetViewQueryTimeout()
+	Infof(KeyAll, "Setting query timeouts for bucket %s to %v", spec.BucketName, timeoutsConfig.QueryTimeout)
+
+	clusterOptions := gocb.ClusterOptions{
+		Authenticator:  authenticator,
+		SecurityConfig: securityConfig,
+		TimeoutsConfig: timeoutsConfig,
+	}
+
+	if spec.KvPoolSize > 0 {
+		// TODO: Equivalent of kvPoolSize in gocb v2?
+	}
+
+	cluster, err := gocb.Connect(connString, clusterOptions)
+	if err != nil {
+		Infof(KeyAuth, "Unable to connect to cluster: %v", err)
+		return nil, err
+	}
+
+	// TODO: Cluster compatibility not exposed, need to make manual /pools/default/ call?
+
+	// Connect to bucket
+	bucket := cluster.Bucket(spec.BucketName)
+	// TODO: identify required services and add to WaitUntilReadyOptions
+	err = bucket.WaitUntilReady(30*time.Second, nil)
+	if err != nil {
+		Warnf("Error waiting for bucket to be ready: %v", err)
+		return nil, err
+	}
+
+	collection := &Collection{
+		Collection: bucket.DefaultCollection(),
+	}
+
+	return collection, nil
+}
+
+type Collection struct {
+	*gocb.Collection            // underlying gocb Collection
+	Spec             BucketSpec // keep a copy of the BucketSpec for DCP usage
+}
+
+// DataStore
+func (c *Collection) GetName() string {
+	return c.Collection.Name()
+}
+
+func (c *Collection) UUID() (string, error) {
+	return "", errors.New("Not implemented")
+}
+func (c *Collection) Close() {
+	return
+}
+
+func (c *Collection) IsSupported(feature sgbucket.DataStoreFeature) bool {
+	return false
+}
+
+// KV store
+
+func (c *Collection) Get(k string, rv interface{}) (cas uint64, err error) {
+	getResult, err := c.Collection.Get(k, nil)
+	if err != nil {
+		return 0, err
+	}
+	err = getResult.Content(rv)
+	return uint64(getResult.Cas()), err
+}
+
+func (c *Collection) GetRaw(k string) (rv []byte, cas uint64, err error) {
+
+	getOptions := &gocb.GetOptions{
+		Transcoder: gocb.NewRawBinaryTranscoder(),
+	}
+	getRawResult, err := c.Collection.Get(k, getOptions)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	err = getRawResult.Content(&rv)
+	return rv, uint64(getRawResult.Cas()), err
+}
+
+func (c *Collection) GetAndTouchRaw(k string, exp uint32) (rv []byte, cas uint64, err error) {
+	getAndTouchOptions := &gocb.GetAndTouchOptions{
+		Transcoder: gocb.NewRawBinaryTranscoder(),
+	}
+	getAndTouchRawResult, err := c.Collection.GetAndTouch(k, expAsDuration(exp), getAndTouchOptions)
```

Comment:
`err` is unused here. Is it possible getAndTouchRawResult is nil in this case and panic on line 161?

Reply:
Agree - have fixed.